### DISCUSSION TITLE
프로세스란?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#271] [Operating system]
PCB,  메모리 영역(데이터 코드 힙 스택..) 프로세스 상태, 계층 구조(fork exec)

**Comment by 3Juhwan** [comment]
## 프로세스
프로세스는 단순하게 '실행 중인 프로그램'를 의미한다. 정확히는 'OS에게 자원을 할당받은 최소 작업 단위'를 의미한다. 

## 프로세스가 할당 받는 자원 
프로세스가 생성되면 메모리의 커널 영역에 PCB를, 사용자 영역에는 힙/스택/데이터/코드 영역을 할당 받는다. 

**PCB (Process Control Block)**
프로세스 제어에 필요한 정보를 저장한다. OS가 프로세스의 실행 순서를 관리하고, 자원을 배분할 때 PCB를 사용한다. 

- PID(Process ID): 프로세스를 식별하기 위해 부여한 번호다. 
- 레지스터 값: 프로세스가 실행되기 위해 이전에 사용했던 레지스터의 중간 값을 복원해야 한다. 
- 프로세스 상태: 현재 프로세스가 어떤 상태인지 기록되어야 한다. 상태는 생성/준비/실행/대기/종료 상태가 있다. 
- CPU 스케줄링 정보: 프로세스가 언제, 어떤 순서로 CPU를 할당받을지에 대한 정보다. 
- 메모리 관리 정보: 프로세스가 메모리에 어떤 위치에 저장되었는지에 대한 정보다. 베이스 레지스터, 한계 레지스터, 페이지 테이블 정보 등이 담긴다. 
- 사용한 파일과 입출력장치 목록: 프로세스가 사용한 파일과 할당받은 입출력장치에 대한 정보다. 

</br>

**메모리(사용자 영역)**
- 코드 영역: 실행할 수 있는 코드, 기계어 명령어를 저장한다. CPU가 실행할 명령어로 **읽기 전용** 영역이다. 
- 데이터 영역: 프로세스를 실행하며 할당되어 종료까지 유지되는 데이터 영역이다. 
- 힙 영역: 사용자에 의해 메모리 공간이 동적으로 할당 및 해제된다. 메모리의 낮은 주소에서 높은 주소의 방향으로 할당한다. 
- 스택 영역: 데이터를 일시적으로 저장하는 공간이다. 함수 실행에 따라 할당되는 매개 변수, 지역 변수가 예시이다. 

## 프로세스 상태
CPU는 빠른 속도로 프로세스를 번갈아 가며 실행한다. 이에 따라 프로세스는 여러 상태를 거치며 실행된다. 운영체제는 PCB를 통해 프로세스 상태를 확인하고 관리한다. 

![image](https://github.com/user-attachments/assets/ffd0d36a-3fc4-4e60-b353-a7fc9cb9a154)


- 생성 상태: 프로세스가 생성 중인 상태로 운영체제에게 PCB를 할당받은 상태다. 생성이 완료되면 준비 상태로 이동한다. 
- 준비 상태: CPU 할당을 기다리는 상태다. 자신의 차례가 오면 실행 상태로 이동한다. 
- 실행 상태: CPU를 할당받아 실행 중인 상태다. CPU는 일정 시간만 사용할 수 있고, 할당받은 시간이 지나면 타이머 인터럽트가 발생하여 다시 준비 상태로 이동한다. 실행 도중 입출력장치를 사용한다면 대기 상태로 이동한다. 
- 대기 상태: 입출력장치의 작업을 기다리는 상태다. 입출력 작업이 완료되면 대기 상태로 이동한다. 
- 종료 상태: 프로세스가 종료된 상태다. 운영체제는 PCB와 사용한 메모리를 정리한다. 

## 프로세스 계층 구조
프로세스는 실행 도중에 시스템 호출을 통해 다른 프로세스를 생성할 수 있다. 기존 프로세스는 부모 프로세스, 새로 생성된 프로세스는 자식 프로세스라고 한다. PCB에 부모 프로세스의 PID인 PPID가 기록되기도 한다. 

- fork(): 자식 프로세스를 생성하는 시스템 호출이다. 부모 프로세스의 자원들을 상속 받지만 저장된 메모리 위치는 다르다. 
- exec(): 자신의 메모리 공간을 다른 프로그램으로 교체하는 시스템 호출이다. 코드 영역과 데이터 영역은 기존 메모리 내용을 덮어쓰고, 힙 영역과 스택 영역은 초기화한다. 

fork()를 실행하고 exec()는 실행하지 않는다면 동일한 프로세스가 병행하여 실행된다.
